pinot-controller/src/main/java/com/linkedin/pinot/controller/api/upload/ZKOperator.java

```diff
@@ -0,0 +1,212 @@
+        // New segment is different with the existing one, update ZK metadata and refresh the segment
+        LOGGER.info("New segment crc {} is different than the existing segment crc {}. Updating ZK metadata and refreshing segment {}",
+            newCrc, existingCrc, segmentName);
+        moveSegmentToPermanentDirectory(currentSegmentLocation, finalSegmentLocationURI);
+        LOGGER.info("Moved segment {} from temp location {} to {}", segmentName, currentSegmentLocation.getAbsolutePath(), finalSegmentLocationURI.getPath());
+        if (finalSegmentLocationURI.getScheme().equals(LOCAL_SEGMENT_SCHEME)) {
+          String downloadUrl = ControllerConf.constructDownloadUrl(rawTableName, segmentName, provider.getVip());
+          _pinotHelixResourceManager.refreshSegment(segmentMetadata, existingSegmentZKMetadata, downloadUrl);
+
+        } else {
+          _pinotHelixResourceManager.refreshSegment(segmentMetadata, existingSegmentZKMetadata, finalSegmentLocationURI.toString());
+        }
+      }
+    } catch (Exception e) {
+      if (!_pinotHelixResourceManager.updateZkMetadata(existingSegmentZKMetadata)) {
+        LOGGER.error("Failed to update ZK metadata for segment: {} of table: {}", segmentName, offlineTableName);
+      }
+      throw e;
+    }
+  }
+
+  private void checkCRC(HttpHeaders headers, String offlineTableName, String segmentName, long existingCrc) {
+    String expectedCrcStr = headers.getHeaderString(HttpHeaders.IF_MATCH);
+    if (expectedCrcStr != null) {
+      long expectedCrc;
+      try {
+        expectedCrc = Long.parseLong(expectedCrcStr);
+      } catch (NumberFormatException e) {
+        throw new ControllerApplicationException(LOGGER,
+            "Caught exception for segment: " + segmentName + " of table: " + offlineTableName
+                + " while parsing IF-MATCH CRC: \"" + expectedCrcStr + "\"", Response.Status.PRECONDITION_FAILED);
+      }
+      if (expectedCrc != existingCrc) {
+        throw new ControllerApplicationException(LOGGER,
+            "For segment: " + segmentName + " of table: " + offlineTableName + ", expected CRC: " + expectedCrc
+                + " does not match existing CRC: " + existingCrc, Response.Status.PRECONDITION_FAILED);
+      }
+    }
+  }
+
+  private void processNewSegment(SegmentMetadata segmentMetadata, URI finalSegmentLocationURI,
+      File currentSegmentLocation, FileUploadPathProvider provider, String rawTableName, String segmentName) {
+    try {
+      moveSegmentToPermanentDirectory(currentSegmentLocation, finalSegmentLocationURI);
+      LOGGER.info("Moved segment {} from temp location {} to {}", segmentName, currentSegmentLocation.getAbsolutePath(), finalSegmentLocationURI.getPath());
+    } catch (Exception e) {
+      LOGGER.error("Could not move segment {} from table {} to permanent directory", segmentName, rawTableName, e);
+      throw new RuntimeException(e);
+    }
+    if (finalSegmentLocationURI.getScheme().equals(LOCAL_SEGMENT_SCHEME)) {
```

Comment:
I think we can get rid of this 'if' by introducing another parameter. Basically, ZKOperator will need the following:
1. Metadata
2. sourceUri: URI of the source
3. downloadURI: URI where servers download segments from
4. storageURI: URI where segments are stored.
3 and 4 can be same in all cases except for segment upload (with data).